stdlib/src/precompiles/keccak.rs

```diff
@@ -0,0 +1,82 @@
+use alloc::{vec, vec::Vec};
+
+use miden_core::{AdviceMap, Felt, crypto::hash::Digest};
+use miden_crypto::hash::{keccak::Keccak256, rpo::Rpo256};
+use miden_processor::{AdviceMutation, EventError, ProcessState};
+
+use crate::precompiles::read_memory;
+
+pub const KECCAK_EVENT_ID: &str = "miden_stdlib::hash::keccak";
+
+// KECCAK EVENT ERROR
+// ================================================================================================
+
+#[derive(Debug, thiserror::Error)]
+pub enum KeccakError {
+    #[error("failed to read memory at ptr {ptr}, len {len}")]
+    MemoryReadFailed { ptr: u64, len: u64 },
+    #[error("invalid byte at index {byte_index} with value: {value}")]
+    InvalidByteValue { value: u64, byte_index: usize },
+}
+
+/// Event handler which pushes keccak256 hash to the advice stack.
+///
+/// This handler is used to defer expensive keccak computation to the native verifier.
+/// The verifier will re-execute the keccak operation and verify the result.
+///
+/// Inputs:
+///   Operand stack: [event_id, ptr, len, ...]
+///   Memory: bytes at [ptr, len)
+///
+/// Outputs:
+///   Advice stack: [hash_bytes...] (32 bytes as individual Felts)
+///   Advice map: commitment -> preimage mapping for prover recovery
+///
+/// Where:
+/// - ptr is the memory address where the input bytes start
+/// - len is the number of bytes to hash
+/// - hash_bytes are the 32 bytes of the keccak256 hash
+pub fn push_keccak(process: &ProcessState) -> Result<Vec<AdviceMutation>, EventError> {
+    // start at 1 since 0 holds the event id
+    // Note: stack layout after emit is [event_id, ptr, len, ...]
+    let ptr = process.get_stack_item(1).as_int();
+    let len = process.get_stack_item(2).as_int();
+
+    // Attempt to collect the field elements between `ptr` and `ptr+len`.
+    let witness =
+        read_memory(process, ptr, len).ok_or(KeccakError::MemoryReadFailed { ptr, len })?;
+
+    // Try to convert to bytes
+    let preimage = witness
+        .iter()
+        .enumerate()
+        .map(|(idx, felt)| {
+            u8::try_from(felt.as_int()).map_err(|_| KeccakError::InvalidByteValue {
+                value: felt.as_int(),
+                byte_index: idx,
+            })
+        })
+        .collect::<Result<Vec<_>, _>>()?;
```

Comment:
Why do we assume that each element contains just 1 byte? I think we can safely assume that each element has 4 bytes. Or does this create some problems?